pkg/agent/openflow/pipeline.go

```diff
@@ -2530,77 +2525,73 @@ func (c *client) genPacketInMeter(meterID binding.MeterIDType, rate uint32) bind
-	c.createOFTable(L2ForwardingOutTable, binding.LastTableID, binding.TableMissActionDrop)
-	if c.enableProxy {
-		SpoofGuardTable.SetNext(ServiceHairpinTable.GetID())
-		IPv6Table.SetNext(ServiceHairpinTable.GetID())
-		ConntrackStateTable.SetNext(EndpointDNATTable.GetID())
-		if c.proxyAll {
-			ServiceHairpinTable = c.createOFTable(ServiceHairpinTable, ServiceConntrackTable.GetID(), binding.TableMissActionNext)
-			ServiceConntrackTable = c.createOFTable(ServiceConntrackTable, ConntrackTable.GetID(), binding.TableMissActionNext)
-			ServiceClassifierTable = c.createOFTable(ServiceClassifierTable, binding.LastTableID, binding.TableMissActionNone)
-			ServiceConntrackCommitTable = c.createOFTable(ServiceConntrackCommitTable, HairpinSNATTable.GetID(), binding.TableMissActionNext)
-		} else {
-			ServiceHairpinTable = c.createOFTable(ServiceHairpinTable, ConntrackTable.GetID(), binding.TableMissActionNext)
-		}
-		SessionAffinityTable = c.createOFTable(SessionAffinityTable, binding.LastTableID, binding.TableMissActionNone)
-		ServiceLBTable = c.createOFTable(ServiceLBTable, EndpointDNATTable.GetID(), binding.TableMissActionNext)
-		EndpointDNATTable = c.createOFTable(EndpointDNATTable, c.egressEntryTable, binding.TableMissActionNext)
-		HairpinSNATTable = c.createOFTable(HairpinSNATTable, L2ForwardingOutTable.GetID(), binding.TableMissActionNext)
-		ConntrackCommitTable.SetNext(HairpinSNATTable.GetID())
-	} else {
-		c.createOFTable(DNATTable, c.egressEntryTable, binding.TableMissActionNext)
-	}
-	if c.enableEgress {
-		c.createOFTable(SNATTable, L2ForwardingCalcTable.GetID(), binding.TableMissActionNext)
-	}
-	if c.enableAntreaPolicy {
-		c.createOFTable(AntreaPolicyEgressRuleTable, EgressRuleTable.GetID(), binding.TableMissActionNext)
-		c.createOFTable(AntreaPolicyIngressRuleTable, IngressRuleTable.GetID(), binding.TableMissActionNext)
-	}
-	if c.enableMulticast {
-		SpoofGuardTable.SetNext(MulticastTable.GetID())
-		c.createOFTable(MulticastTable, ConntrackTable.GetID(), binding.TableMissActionNext)
-		if c.enableProxy {
-			MulticastTable.SetNext(ServiceHairpinTable.GetID())
+func generatePipeline(pipelineID binding.PipelineID, requiredTables []*Table) binding.Pipeline {
+	var ofTables []binding.Table
+	for _, table := range requiredTables {
+		// Generate a sequencing ID for the flow table.
+		tableID := binding.NextTableID()
+		// Initialize a flow table.
+		table.ofTable = binding.NewOFTable(tableID, table.name, table.stage, pipelineID)
+		ofTables = append(ofTables, table.ofTable)
+		tableCache.Add(table)
+	}
+	return binding.NewPipeline(pipelineID, ofTables)
+}
+
+// realizePipelines sets nextID and missing action for every flow table in every pipeline and realize it on OVS bridge.
+func (c *client) realizePipelines() {
+	tableMissDropTables := map[string]struct{}{
+		PipelineRootClassifierTable.GetName(): {},
+		ClassifierTable.GetName():             {},
+		SpoofGuardTable.GetName():             {},
+	}
```

Comment:
Better not to have business specific code in a generic function, the point of the framework is decouple the generic part and business part. I think if a table is supposed to drop packets if they don't match any of its flows, it should be declared when creating the table.